src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/GroupCollection.cs

```diff
@@ -205,7 +205,55 @@ void IList.RemoveAt(int index)
             set { throw new NotSupportedException(SR.NotSupported_ReadOnlyCollection); }
         }
 
-        private sealed class Enumerator : IEnumerator<Group>
+        IEnumerator<KeyValuePair<string, Group>> IEnumerable<KeyValuePair<string, Group>>.GetEnumerator()
+        {
+            return new Enumerator(this);
+        }
+
+        public bool TryGetValue(string key, out Group value)
+        {
+            var group = this[key];
+            if (group == Group.s_emptyGroup)
+            {
+                value = null;
+                return false;
+            }
+            value = group;
+            return true;
+        }
+
+        public bool ContainsKey(string key)
+        {
+            return _match._regex.GroupNumberFromName(key) >= 0;
+        }
+
+        public IEnumerable<string> Keys
+        {
+            get
+            {
+                var enumerator = new Enumerator(this);
+
+                while (enumerator.MoveNext())
+                {
+                    yield return enumerator.Current.Name;
+                }
+            }
+        }
```

Comment:
This has the same issue I previously mentioned, that there are two allocations here.  The first allocation is for the Enumerable returned from Keys, which the compiler will then also use as the enumerator returned from its GetEnumerator call.  The second allocation is for the `new Enumerator(this)` above.  You don't need both.  Can't this instead just be implemented simply like:
```C#
public IEnumerable<Group> Keys
{
    for (int i = 0; i < Count; i++)
    {
        yield return this[i].Name;
    }
}
```
and similarly for Values?